core/src/main/java/org/elasticsearch/tasks/CancellableTask.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.tasks;
+
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.inject.Provider;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A task that can be canceled
+ */
+public class CancellableTask extends Task {
+
+    private final boolean hasChildren;
+
+    private static final Tuple<String, CancellationListener> TASK_FINISHED_MARKER = new Tuple<>("task finished", null);
+
+    private static ThreadLocal<CancellableTask> current = new ThreadLocal<>();
+
+    public static void setCurrent(CancellableTask value) {
+        current.set(value);
+    }
+
+    public static void removeCurrent() {
+        current.remove();
+    }
+
+    public static void checkCurrent() {
+        current.get().check();
+    }
+
+    private AtomicReference<Tuple<String, CancellationListener>> cancellationReasonAndListener = new AtomicReference<>();
+
+    public CancellableTask(long id, String type, String action, Provider<String> description, boolean hasChildren) {
+        super(id, type, action, description);
+        this.hasChildren = hasChildren;
+    }
+
+    public CancellableTask(long id, String type, String action, Provider<String> description, String parentNode, long parentId, boolean hasChildren) {
+        super(id, type, action, description, parentNode, parentId);
+        this.hasChildren = hasChildren;
+    }
+
+    public boolean hasChildren() {
+        return hasChildren;
+    }
+
+    public final boolean cancel(String reason, CancellationListener listener) {
```

Comment:
I don't think we should keep any state in here at all and instead move everything we keep on state in the TaskManager. That way we don't need the listener on this class and also maybe we can skip it entirely if we keep state on the TaskManager what has children and what is cancelable?